aptos-move/aptos-vm/src/block_executor/vm_wrapper.rs

```diff
@@ -45,15 +45,10 @@ impl<'a, S: 'a + StateView + Sync> ExecutorTask for AptosExecutorTask<'a, S> {
         txn_idx: TxnIndex,
         materialize_deltas: bool,
     ) -> ExecutionStatus<AptosTransactionOutput, VMStatus> {
-        // TODO[agg_v2](fix) look at whether it is enabled, not just capable.
-        // if it is disabled, no need to fallback.
-        if txn.is_valid() && executor_with_group_view.is_delayed_field_optimization_capable() {
-            if let Transaction::GenesisTransaction(WriteSetPayload::Direct(_)) = txn.expect_valid()
-            {
-                // WriteSetPayload::Direct cannot be handled in mode where delayed_field_optimization is enabled
-                // And we need to communicate to the BlockExecutor, so they can retry with capability disabled
-                return ExecutionStatus::DirectWriteSetTransactionNotCapableError;
-            }
+        if executor_with_group_view.is_delayed_field_optimization_capable()
+            || executor_with_group_view.is_resource_group_split_in_change_set_capable()
+        {
+            assert!(Self::is_transaction_dynamic_change_set_capable(txn));
```

Comment:
nit: Looks like the invariant we want to check here is that if the transaction is not capable, views aren't advertising being capable, so should we structure the if / assert in a different order?

Reply:
not x => not y  

is the same as

y => x

:)


I am logically saying here - if we are in the capable context, we cannot be processing transactions that cannot be handled that way - seems reasonable order to me